src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/ModelExplorer.cs

```diff
@@ -0,0 +1,378 @@
+        /// <see cref="ModelExplorer"/> was provided with one.
+        /// </remarks>
+        public object Model
+        {
+            get
+            {
+                if (_model == null && _modelAccessor != null)
+                {
+                    Debug.Assert(Container != null);
+                    _model = _modelAccessor(Container.Model);
+
+                    // Null-out the accessor so we don't invoke it repeatedly if it returns null.
+                    _modelAccessor = null;
+                }
+
+                return _model;
+            }
+        }
+
+        /// <remarks>
+        /// Retrieving the <see cref="Model"/> type will for execution of the model accessor function if this
+        /// <see cref="ModelExplorer"/> was provided with one.
+        /// </remarks>
+        public Type ModelType
+        {
+            get
+            {
+                if (_modelType == null)
+                {
+                    if (Model == null)
+                    {
+                        // If the model is null, then use the declared model type;
+                        _modelType = Metadata.ModelType;
+                    }
+                    else if (Metadata.IsNullableValueType)
+                    {
+                        // We have a model, but if it's a nullable value type, then Model.GetType() will return
+                        // the non-nullable type (int? -> int). Since it's a value type, there's no subclassing,
+                        // just go with the declared type.
+                        _modelType = Metadata.ModelType;
+                    }
+                    else
+                    {
+                        // We have a model, and it's not a nullable so use the runtime type to handle
+                        // cases where the model is a subclass of the declared type and has extra data.
+                        _modelType = Model.GetType();
+                    }
+                }
+
+                return _modelType;
+            }
+        }
+
+        /// <summary>
+        /// Gets the properties.
+        /// </summary>
+        /// <remarks>
+        /// Includes a <see cref="ModelExplorer"/> for each property of the <see cref="ModelMetadata"/>
+        /// for <see cref="ModelType"/>. 
+        /// </remarks>
+        public IEnumerable<ModelExplorer> Properties
+        {
+            get
+            {
+                if (_properties == null)
+                {
+                    _properties = new List<ModelExplorer>();
+
+                    var metadata = GetMetadataForRuntimeType();
+
+                    var properties = Enumerable.Join(
+                        metadata.Properties,
+                        PropertyHelper.GetProperties(ModelType),
+                        m => m.PropertyName,
+                        ph => ph.Property.Name,
+                        (m, ph) => new { Metadata = m, PropertyHelper = ph });
```

Comment:
nits:
- rather than the anonymous object, return `CreateExplorerForProperty(property.Metadata, property.PropertyHelper)`
- elongate the lambda parameter names
- might need a comment like `// Find the ModelMetadata and PropertyHelper pairs for each property.`